File: app/scripts/tablero.lua
---
-- Created by IntelliJ IDEA.
-- User: mcadavip
-- Date: 31/10/11
-- Time: 02:55 PM
-- To change this template use File | Settings | File Templates.
--

tablero = {}
tablero.__index = tablero

function tablero.crear()
  local tab = {}             -- el nuevo objeto
  setmetatable(tab,tablero)  -- hace una ubicación del tablero

  tab.tamanocuadro = 10
  tab.LIMITE_ABAJO = 31
  tab.LIMITE_IZQUIERDO = -1
  tab.LIMITE_DERECHO = 16
  tab.LIMITE_ARRIBA = -5

  tab.matriz = {} -- create the matrix

  for i=0,15 do
    tab.matriz[i] = {} -- create a new row
    for j=0,30 do
      tab.matriz[i][j] = nil
    end
  end

  return tab
end

function tablero:PuedeRotar(cuadros)
  local i = 0
  while i < 4 do
    if cuadros[i].ubicacion.X == -1 or cuadros[i].ubicacion.X == 16 or
      cuadros[i].ubicacion.Y == -1 or cuadros[i].ubicacion.Y == 31 or
      self.matriz[cuadros[i].ubicacion.X][cuadros[i].ubicacion.Y] ~= nil then
      return false
    end
    i = i + 1
  end
  return true;
end

function tablero:PuedeDesplazarse(tipomovimiento, cuadros)
  if tipomovimiento.TIPO == "IZQUIERDA" then
    local i = 0
    while i < 4 do
      if cuadros[i].ubicacion.X == 0 or self.matriz[cuadros[i].ubicacion.X - 1][cuadros[i].ubicacion.Y] ~= nil then
        return false
      end
      i = i + 1
    end

    else if tipomovimiento.TIPO == "DERECHA" then
      local i = 0
      while i < 4 do
        if cuadros[i].ubicacion.X == 15 or self.matriz[cuadros[i].ubicacion.X + 1][cuadros[i].ubicacion.Y] ~= nil then
          return false
        end
        i = i + 1
      end

      else if tipomovimiento.TIPO == "ABAJO" then
        local i = 0
        while i < 4 do
          if cuadros[i].ubicacion.Y == 30 or self.matriz[cuadros[i].ubicacion.X][cuadros[i].ubicacion.Y + 1] ~= nil then
            return false
          end
          i = i + 1
        end
      end
    end
  end
  return true;
end

function tablero:IngresarCuadradosMatriz(cuadros)
  local i = 0
  while i < 4 do
    self.matriz[cuadros[i].ubicacion.X][cuadros[i].ubicacion.Y] = cuadros[i]
    i = i + 1
  end
end

function tablero:EliminarCuadradoMatriz(cuadro)
  self.matriz[cuadro.ubicacion.X][cuadro.ubicacion.Y] = nil;
end

function tablero:ConsultarFilaLlena()
  local filallena = 0

  for y = self.LIMITE_ABAJO - 1, 0, -1 do
    filallena = 0;
    for x = self.LIMITE_IZQUIERDO + 1, self.LIMITE_DERECHO -1 do
      if self.matriz[x][y] ~= nil then
          filallena = filallena + 1
      end
    end

    if filallena == 16 then
        return y;
    end
  end

  return -1;
end

function tablero:EliminarFilaLlena(fila)
  for y = fila - 1, 1, -1 do
    for x = self.LIMITE_IZQUIERDO + 1, self.LIMITE_DERECHO - 1 do
      self.matriz[x][fila] = self.matriz[x][y]

      if self.matriz[x][fila] ~= nil then
        self.matriz[x][fila]:MoverCuadradoAbajo()
      end
    end
    fila = fila-1
  end
  self:Redibujar();
end

function tablero:Redibujar()
  local color = display.newRect(0,0,display.contentWidth,display.contentHeight)
  color:setFillColor(255,255,255)

  for i=0,15 do
    for j=0,30 do
      if self.matriz[i][j] ~= nil then
        self.matriz[i][j]:mostrar()
      end
    end
  end
end

File: app/scripts/tablero.js
TETRIS.tablero = function () {
	this.tamanocuadro;
	this.LIMITE_ABAJO;
	this.LIMITE_IZQUIERDO;
	this.LIMITE_DERECHO;
	this.LIMITE_ARRIBA;

	this.matriz = {};

	this.crear = function () {
		this.tamanocuadro = 10;
		this.LIMITE_ABAJO = 31;
		this.LIMITE_IZQUIERDO = -1;
		this.LIMITE_DERECHO = 16;
		this.LIMITE_ARRIBA = -5;

		this.matriz = {};

		for (i=0; i<10; i++){
			this.matriz[i] = {};
		
			for (j=0; j<20; j++){
				this.matriz[i][j] = null;
			}
		}	
	}

	this.puedeRotar = function(cuadros){
		i = 0;
		
		while (i < 4) {
			if (cuadros[i].ubicacion.X == -1 || cuadros[i].ubicacion.X == 10 ||
				cuadros[i].ubicacion.Y == -1 || cuadros[i].ubicacion.Y == 20 ||
				this.matriz[cuadros[i].ubicacion.X][cuadros[i].ubicacion.Y] != null) {
				return false
			}
			i = i + 1;
		}
		return true;
	}

	this.puedeDesplazarse = function(tipomovimiento, cuadros){
		if (tipomovimiento == "IZQUIERDA") {
			i = 0;
			
			while (i < 4) {
				if (cuadros[i].ubicacion.X == 0 || this.matriz[cuadros[i].ubicacion.X - 1][cuadros[i].ubicacion.Y] != null) {
					return false;
				}
			
				i = i + 1;
			}
		}
		else if (tipomovimiento == "DERECHA") {
			i = 0;
			
			while (i < 4) {
				if (cuadros[i].ubicacion.X == 9 || this.matriz[cuadros[i].ubicacion.X + 1][cuadros[i].ubicacion.Y]!= null) {
					return false;
				}
				i = i + 1;
			}
		}
		else if (tipomovimiento == "ABAJO") {
			i = 0;
			
			while (i < 4) {
				if (cuadros[i].ubicacion.Y == 19 || this.matriz[cuadros[i].ubicacion.X][cuadros[i].ubicacion.Y + 1]!= null) {
					return false;
				}
				
				i = i + 1;
			}
		}
		return true;
	}

	this.ingresarCuadradosMatriz = function(cuadros){
		i = 0;
		
		while (i < 4) {
			this.matriz[cuadros[i].ubicacion.X][cuadros[i].ubicacion.Y] = cuadros[i];
			i = i + 1;
		}
	}

	this.eliminarCuadradoMatriz = function(cuadro){
  		this.matriz[cuadro.ubicacion.X][cuadro.ubicacion.Y] = null;
	}

	this.consultarFilaLlena = function(){
		filallena = 0;

		for (y = this.LIMITE_ABAJO - 1; y > 0; y--) {
			filallena = 0;
			
			for (x = this.LIMITE_IZQUIERDO + 1; y > this.LIMITE_DERECHO; y--) {
				if (this.matriz[x][y] != null) {
					filallena = filallena + 1;
				}
			}

			if (filallena == 21) {
				return y;
			}
		}

		return -1;
	}

	this.eliminarFilaLlena = function(fila){
		for (y = fila - 1; y > 1; y--) {
			for (x = this.LIMITE_IZQUIERDO + 1; x > this.LIMITE_DERECHO; x--) {
				this.matriz[x][fila] = this.matriz[x][y];

				if (this.matriz[x][fila]!= null) {
					this.matriz[x][fila].moverCuadradoAbajo();
				}
			}
			fila = fila-1;
		}
		this.redibujar();
	}

	this.redibujar = function(){
		//color = display.newRect(0,0,display.contentWidth,display.contentHeight)
		//color:setFillColor(255,255,255)

		for (i=0; i < 10; i++) {
			for (j=0; j < 20; j++) {
				if (this.matriz[i][j]!= null) {
					this.matriz[i][j].mostrar();
				}
			}
		}
	}
};